core/src/main/java/org/apache/iceberg/view/ViewMetadataParser.java

```diff
@@ -0,0 +1,213 @@
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      generator.useDefaultPrettyPrinter();
+      toJson(metadata, generator);
+      generator.flush();
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to write json to file: %s", outputFile);
+    }
+  }
+
+  public static void toJson(ViewMetadata metadata, JsonGenerator generator) throws IOException {
+    generator.writeStartObject();
+
+    generator.writeNumberField(FORMAT_VERSION, metadata.formatVersion());
+    generator.writeStringField(LOCATION, metadata.location());
+
+    JsonUtil.writeStringMap(PROPERTIES, metadata.properties(), generator);
+
+    if (metadata.schemas() != null && !metadata.schemas().isEmpty()) {
+      generator.writeNumberField(CURRENT_SCHEMA_ID, metadata.currentSchemaId());
+      generator.writeArrayFieldStart(SCHEMAS);
+      for (Schema schema : metadata.schemas()) {
+        SchemaParser.toJson(schema, generator);
+      }
+
+      generator.writeEndArray();
+    }
+
+    generator.writeNumberField(CURRENT_VERSION_ID, metadata.currentVersionId());
+    generator.writeArrayFieldStart(VERSIONS);
+    for (ViewVersion version : metadata.versions()) {
+      ViewVersionParser.toJson(version, generator);
+    }
+
+    generator.writeEndArray();
+
+    generator.writeArrayFieldStart(VERSION_LOG);
+    for (ViewHistoryEntry viewHistoryEntry : metadata.history()) {
+      ViewHistoryEntryParser.toJson(viewHistoryEntry, generator);
+    }
+    generator.writeEndArray();
+
+    generator.writeEndObject();
+  }
+
+  static String toJson(ViewMetadata viewMetadata) {
+    return JsonUtil.generate(gen -> toJson(viewMetadata, gen), false);
+  }
+
+  public static ViewMetadata read(InputFile file) {
+    try (InputStream is = file.newStream()) {
+      return fromJson(JsonUtil.mapper().readValue(is, JsonNode.class));
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to read file: %s", file);
+    }
+  }
+
+  public static ViewMetadata fromJson(JsonNode node) {
+    Preconditions.checkArgument(node != null, "Cannot parse view metadata from null json");
+    Preconditions.checkArgument(
+        node.isObject(), "Cannot parse metadata from a non-object: %s", node);
+
+    int formatVersion = JsonUtil.getInt(FORMAT_VERSION, node);
+    Preconditions.checkArgument(
+        formatVersion <= ViewMetadata.SUPPORTED_VIEW_FORMAT_VERSION,
+        "Cannot read unsupported version %s",
+        formatVersion);
+
+    String location = JsonUtil.getString(LOCATION, node);
+
+    int currentVersionId = JsonUtil.getInt(CURRENT_VERSION_ID, node);
+
+    Map<String, String> properties = JsonUtil.getStringMap(PROPERTIES, node);
+    JsonNode versionsListNode = JsonUtil.get(VERSIONS, node);
+    List<ViewVersion> versions = Lists.newArrayListWithExpectedSize(versionsListNode.size());
+    Iterator<JsonNode> iterator = versionsListNode.elements();
```

Comment:
should we validate the list type for the JSON node? this applies to other non-primitive types (like Map).
```
Preconditions.checkArgument(
          versionsListNode.isArray(), "Cannot parse ... from a non-array: %s", versionsListNode);
````

If we define the schema for `ViewMetadata`, we might be able to leverage `SingleValueParser` to handle those. E.g.
```
    Map<Integer, ByteBuffer> upperBounds = null;
    if (jsonNode.has(UPPER_BOUNDS)) {
      upperBounds =
          (Map<Integer, ByteBuffer>)
              SingleValueParser.fromJson(DataFile.UPPER_BOUNDS.type(), jsonNode.get(UPPER_BOUNDS));
    }
```

Reply:
+1. We always check the types to give clear error messages.

Reply:
Good catch, I updated to verify the data types so the error message is obvious